---
title: "HiC-Compartment_analysis_Mastropasqua_et_al_22"
author: "F Mastropasqua"
date: "10/4/2022"
output: html_document
---

```{r}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
library(dplyr)
library(ggplot2)
library(BiocManager)
```

##Generate composite eigenvector for somatic chromosomes

Load bedgraph files of the PCAs generated by hicPCA command from Hic Explorer. The PCAs1-4 have been visually inspected in order to get the PCA that mostly fit with gene density. For all the samples and chromosomes PCA1 corresponded to the p and q arms of the chromosomes. Chromosome 19 was excluded since the PCA calls were not clear. The "ASD" samples correspond to HNRNPUdel/+ samples.

#Composite EV D0

```{r}
####HNRNPU del/+ D0####
ASD_EV_2 <-read.delim("./HiC_files/PCA2_ASD12_D0_ICE_1M_genes.bedgraph", header=F)
ASD_EV_3 <-read.delim("./HiC_files/PCA3_ASD12_D0_ICE_1M_genes.bedgraph", header=F)

#select the EV for each chromosome and merge them
c <- ASD_EV_2[!(ASD_EV_2$V1=="chr1"|ASD_EV_2$V1=="chr16"|ASD_EV_2$V1=="chr19"|ASD_EV_2$V1=="chr22"|ASD_EV_2$V1=="chrX"|ASD_EV_2$V1=="chrY"),]
c1 <-ASD_EV_3[ASD_EV_3$V1=="chr1",]
c16 <-ASD_EV_3[ASD_EV_3$V1=="chr16",]
c22 <-ASD_EV_3[ASD_EV_3$V1=="chr22",]

ASD_COMPOSITE_D0 <- bind_rows(c,c1,c16,c22, .id=NULL )

#add a new A/B column
ASD_COMPOSITE_D0$AB[ASD_COMPOSITE_D0$V4 >= 0] <- "A" 
ASD_COMPOSITE_D0$AB[ASD_COMPOSITE_D0$V4 < 0] <- "B"
#change order of the columns so the EV values are in the column 5
ASD_COMPOSITE_D0 <- ASD_COMPOSITE_D0[, c(1,2,3,5,4)]
colnames(ASD_COMPOSITE_D0) <- c('chrom', 'chromStart', 'chromEnd', 'AB', 'PCA')

####CTRL D0####
CTRL_EV_2 <-read.delim("./HiC_files/PCA2_CTRL9_D0_ICE_1M_genes.bedgraph", header=F)
CTRL_EV_3 <-read.delim("./HiC_files/PCA3_CTRL9_D0_ICE_1M_genes.bedgraph", header=F)


#select the EV for each chromosome and merge them
c <- CTRL_EV_2[!(CTRL_EV_2$V1=="chr1"|CTRL_EV_2$V1=="chr16"|CTRL_EV_2$V1=="chr19"|CTRL_EV_2$V1=="chrX"|CTRL_EV_2$V1=="chrY"),]
c1 <-CTRL_EV_3[CTRL_EV_3$V1=="chr1",]
c16 <-CTRL_EV_3[CTRL_EV_3$V1=="chr16",]

CTRL_COMPOSITE_D0 <- bind_rows(c,c1,c16, .id=NULL )

#add a new A/B column
CTRL_COMPOSITE_D0$AB[CTRL_COMPOSITE_D0$V4 >= 0] <- "A" 
CTRL_COMPOSITE_D0$AB[CTRL_COMPOSITE_D0$V4 < 0] <- "B"
#change order of the columns so the EV values are in the column 5
CTRL_COMPOSITE_D0 <- CTRL_COMPOSITE_D0[, c(1,2,3,5,4)]
colnames(CTRL_COMPOSITE_D0) <- c('chrom', 'chromStart', 'chromEnd', 'AB', 'PCA')
```

#Composite EV D28

```{r}
####HNRNPUdel/+ D28####
ASD_EV_2 <-read.delim("./HiC_files/PCA2_ASD12_D28_ICE_1M_genes.bedgraph", header=F)
ASD_EV_4 <-read.delim("./HiC_files/PCA4_ASD12_D28_ICE_1M_genes.bedgraph", header=F)
ASD_EV_3 <-read.delim("./HiC_files/PCA3_ASD12_D28_ICE_1M_genes.bedgraph", header=F)

#select the EV for each chromosome and merge them
c1 <- ASD_EV_2[!(ASD_EV_2$V1=="chr4"|ASD_EV_2$V1=="chr16"|ASD_EV_2$V1=="chr19"|ASD_EV_2$V1=="chr22"|ASD_EV_2$V1=="chrX"|ASD_EV_2$V1=="chrY"),]
c4 <-ASD_EV_3[ASD_EV_3$V1=="chr4",]
c16 <-ASD_EV_4[ASD_EV_4$V1=="chr16",]
c22 <-ASD_EV_3[ASD_EV_3$V1=="chr22",]

ASD_COMPOSITE_D28 <- bind_rows(c1,c4,c16,c22, .id=NULL )

#add a new A/B column
ASD_COMPOSITE_D28$AB[ASD_COMPOSITE_D28$V4 >= 0] <- "A" 
ASD_COMPOSITE_D28$AB[ASD_COMPOSITE_D28$V4 < 0] <- "B"
#change order of the columns so the EV values are in the column 5 (in order to get a bed file format)
ASD_COMPOSITE_D28 <- ASD_COMPOSITE_D28[, c(1,2,3,5,4)]

####CTRL D28####
CTRL_EV_2 <-read.delim("./HiC_files/PCA2_CTRL9_D28_ICE_1M_genes.bedgraph", header=F)
CTRL_EV_3 <-read.delim("./HiC_files/PCA3_CTRL9_D28_ICE_1M_genes.bedgraph", header=F)
CTRL_EV_4 <-read.delim("./HiC_files/PCA4_CTRL9_D28_ICE_1M_genes.bedgraph", header=F)

#select the EV for each chromosome and merge them
c1 <- CTRL_EV_2[!(CTRL_EV_2$V1=="chr10"|CTRL_EV_2$V1=="chr16"|CTRL_EV_2$V1=="chr19"|CTRL_EV_2$V1=="chr21"|CTRL_EV_2$V1=="chrX"|CTRL_EV_2$V1=="chrY"),]
c10 <-CTRL_EV_3[CTRL_EV_3$V1=="chr10",]
c16 <-CTRL_EV_4[CTRL_EV_4$V1=="chr16",]
c21 <-CTRL_EV_3[CTRL_EV_3$V1=="chr21",]

CTRL_COMPOSITE_D28 <- bind_rows(c1,c10,c16,c21, .id=NULL )

#add a new A/B column
CTRL_COMPOSITE_D28$AB[CTRL_COMPOSITE_D28$V4 >= 0] <- "A" 
CTRL_COMPOSITE_D28$AB[CTRL_COMPOSITE_D28$V4 < 0] <- "B"
#change order of the columns so the EV values are in the column 5
CTRL_COMPOSITE_D28 <- CTRL_COMPOSITE_D28[, c(1,2,3,5,4)]

```

##Compartment Analysis
```{r}
source("./Find_equal.R")
```

D0

```{r}
Compartments_D0 <- Find_equal(CTRL_COMPOSITE_c_D0$AB, ASD_COMPOSITE_c_D0$AB)
Compartments_D0
Compartments_D0$compartment<- factor(Compartments_D0$compartment, c("A","B","AB","BA"))  

#plot compartment composition (Figure 4D in the paper)
bar<- ggplot(Compartments_D0, aes(x="", y=Freq, fill=compartment))+
  geom_bar(stat="identity", width=1, color="white")+
  coord_polar("y", start=0)+
  ggtitle("Compartment switch D0")+
  theme_void() 
bar
write.table(Compartments_D0, "Compartments_switch_D0_R.txt",dec=".", quote=F)

ASD_comp <-table(ASD_COMPOSITE_c_D0$AB)
ASD_comp<-as.data.frame(ASD_comp)
write.table(ASD_comp, "Compartments_ASD_D0_R.txt",dec=".", quote=F)
CTRL_comp <-table(CTRL_COMPOSITE_c_D0$AB)
CTRL_comp <- as.data.frame(CTRL_comp)
write.table(CTRL_comp, "Compartments_CTRL_D0_R.txt",dec=".", quote=F)
```

D28

```{r}
Compartments_D28<- Find_equal(CTRL_COMPOSITE_D28$AB, ASD_COMPOSITE_D28$AB)
Compartments_D28
Compartments_D28$compartment<- factor(Compartments_D28$compartment, c("A","B","AB","BA")) 

#plot compartment composition (Figure 4D in the paper)
bar<- ggplot(Compartments_D28, aes(x="", y=Freq, fill=compartment))+
  geom_bar(stat="identity", width=1, color="white")+
  coord_polar("y", start=0)+
  ggtitle("Compartment switch D28")+
    theme_void() 
bar
write.table(Compartments_D28, "Compartments_switch_D28_R.txt",dec=".", quote=F)

ASD_A <-ASD_COMPOSITE_D28[ASD_COMPOSITE_D28$AB=="A",]
ASD_B <-ASD_COMPOSITE_D28[ASD_COMPOSITE_D28$AB=="B",]
CTRL_A <-CTRL_COMPOSITE_D28[CTRL_COMPOSITE_D28$AB=="A",]
CTRL_B <-CTRL_COMPOSITE_D28[CTRL_COMPOSITE_D28$AB=="B",]

ASD_comp <-table(ASD_COMPOSITE_D28$AB)
ASD_comp<-as.data.frame(ASD_comp)
write.table(ASD_comp, "Compartments_ASD_D28_R.txt",dec=".", quote=F)
CTRL_comp <-table(CTRL_COMPOSITE_D28$AB)
CTRL_comp <- as.data.frame(CTRL_comp)
write.table(CTRL_comp, "Compartments_CTRL_D28_R.txt",dec=".", quote=F)
```

```{r}
#create one df with the composite EV of each sample at D28
CTR_ASD_comp_D28 <- left_join(CTRL_COMPOSITE_D28,ASD_COMPOSITE_D28, by = c('chrom', 'chromStart', 'chromEnd'))

#compare total switch, A to B, B to A, and regions that have the same type of compartment in the two conditions and make them in bed format

switch_D28 <-CTR_ASD_comp_D28[CTR_ASD_comp_D28$AB.x!=CTR_ASD_comp_D28$AB.y,]
switch_AtoB_D28 <-CTR_ASD_comp_D28[CTR_ASD_comp_D28$AB.x=="A" & CTR_ASD_comp_D28$AB.y=="B",]
switch_BtoA_D28 <-CTR_ASD_comp_D28[CTR_ASD_comp_D28$AB.x=="B" & CTR_ASD_comp_D28$AB.y=="A",]
no_switch_D28 <-CTR_ASD_comp_D28[CTR_ASD_comp_D28$AB.x == CTR_ASD_comp_D28$AB.y,]

switch_D28_bed <-switch_D28[,c(1,2,3,4)]
AtoB_D28_bed<- switch_AtoB_D28[,c(1,2,3,4)]
BtoA_D28_bed<- switch_BtoA_D28[,c(1,2,3,4)]
no_switch_D28_bed<-no_switch_D28 [,c(1,2,3,4)]

write.table(switch_D28_bed, file="switch_regions_D28.bed", dec=".", sep = '\t', quote=FALSE, col.names = FALSE, row.names = FALSE)
write.table(AtoB_D28_bed, file="AtoB_D28.bed", dec=".", sep = '\t', quote=FALSE, col.names = FALSE, row.names = FALSE)
write.table(BtoA_D28_bed, file="BtoA_D28.bed", dec=".", sep = '\t', quote=FALSE, col.names = FALSE, row.names = FALSE)
write.table(no_switch_D28_bed, file="no_switch_D28.bed", dec=".", sep = '\t', quote=FALSE, col.names = FALSE, row.names = FALSE)


```

Generate annotated files and visualize the distribution of the genomic regions in the compartments subject to switch

```{r}
BiocManager::install("annotatr")
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
install.packages("devtools")
library(annotatr)
## load the library
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
## list the contents that are loaded into memory
ls('package:TxDb.Hsapiens.UCSC.hg19.knownGene')
## show the db object that is loaded by calling it's name
TxDb.Hsapiens.UCSC.hg19.knownGene
# Select annotations for intersection with regions
annots = c('hg19_basicgenes')

# Build the annotations (a single GRanges object)
annotations = build_annotations(genome = 'hg19', annotations = annots)

library(regioneR)

CTRL9_D28_GR <- toGRanges(CTRL_COMPOSITE_D28)
ASD12_D28_GR <- toGRanges(ASD_COMPOSITE_D28)
switch_D28_GR<- toGRanges(switch_D28_bed)
AtoB_D28_GR<- toGRanges(AtoB_D28_bed)
BtoA_D28_GR<- toGRanges(BtoA_D28_bed)
noswitch_GR<- toGRanges(no_switch_D28_bed)

# Intersect the regions we read in with the annotations
noswitch_D28_annotated = annotate_regions(
  regions = noswitch_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(noswitch_D28_annotated)
df_noswitch_D28_annotated = data.frame(noswitch_D28_annotated)

write.table(df_noswitch_D28_annotated, file="noswitch_D28_annotated_compartments", quote=FALSE, row.names = FALSE)


CTRL_D28_annotated = annotate_regions(
  regions = CTRL9_D28_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(CTRL_D28_annotated)
df_CTRL_D28_annotated = data.frame(CTRL_D28_annotated)

write.table(df_CTRL_D28_annotated, file="CTRL9_D28_annotated_compartments", quote=FALSE, row.names = FALSE)

ASD12_D28_annotated = annotate_regions(
  regions = ASD12_D28_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(ASD12_D28_annotated)
df_ASD12_D28_annotated = data.frame(ASD12_D28_annotated)

write.table(df_ASD12_D28_annotated, file="ASD12_D28_annotated_compartments", quote=FALSE, row.names = FALSE)

switch_annotated = annotate_regions(
  regions = switch_D28_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(switch_annotated)
df_switch_annotated = data.frame(switch_annotated)

write.table(df_switch_annotated, file="all_switch_D28_annotated", quote=FALSE, row.names = FALSE)


AtoB_annotated = annotate_regions(
  regions = AtoB_D28_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(AtoB_annotated)
df_AtoB_annotated = data.frame(AtoB_annotated)

BtoA_annotated = annotate_regions(
  regions = BtoA_D28_GR,
  annotations = annotations,
  ignore.strand = TRUE,
  quiet = FALSE)
print(BtoA_annotated)
df_BtoA_annotated = data.frame(BtoA_annotated)
# View the number of regions per annotation. This function
annots_order = c(
  'hg19_genes_1to5kb',
  'hg19_genes_promoters',
  'hg19_genes_5UTRs',
  'hg19_genes_exons',
  'hg19_genes_intronexonboundaries',
  'hg19_genes_introns',
  'hg19_genes_3UTRs',
  'hg19_genes_intergenic')


typesC28 <- table(df_CTRL_D28_annotated["annot.type"])
write.table(typesC28, file="CTRL9_D28_annot_type", sep = '\t', quote=FALSE, row.names = FALSE)

typesA28 <- table(df_ASD12_D28_annotated["annot.type"])
write.table(typesA28, file="ASD12_D28_annot_type", sep = '\t', quote=FALSE, row.names = FALSE)

types <- table(df_switch_annotated["annot.type"])
write.table(types, file="switch_D28_annot_type", sep = '\t', quote=FALSE, row.names = FALSE)

types <-read.delim("switch_D28_annot_type")

ggplot(types,aes(x = Var1, y = Freq))+ 
  geom_bar(stat="identity", position = "dodge")+
  ggtitle('# of Sites on all switch compartments at D28')+
  theme_classic()

typesAB <- table(df_AtoB_annotated["annot.type"])
write.table(typesAB, file="AtoB_switch_D28_annot_type", sep = '\t', quote=FALSE, row.names = FALSE)

typesAB <-read.delim("AtoB_switch_D28_annot_type")

ggplot(typesAB,aes(x = Var1, y = Freq))+ 
  geom_bar(stat="identity", position = "dodge")+
  ggtitle('# of  Sites on A to B switch compartments at D28')+
  theme_classic()

typesBA <- table(df_BtoA_annotated["annot.type"])
write.table(typesBA, file="BtoA_switch_D28_annot_type", sep = '\t', quote=FALSE, row.names = FALSE)

typesBA <-read.delim("BtoA_switch_D28_annot_type")

ggplot(typesBA,aes(x = Var1, y = Freq))+ 
  geom_bar(stat="identity", position = "dodge")+
  ggtitle('# of  Sites on B to A switch compartments at D28')+
  theme_classic()

#select A and B compartments in each condition
df_A_CTRL_D28 <-df_CTRL_D28_annotated[df_CTRL_D28_annotated$AB=="A",]
df_B_CTRL_D28 <-df_CTRL_D28_annotated[df_CTRL_D28_annotated$AB=="B",]

df_A_ASD_D28 <-df_ASD12_D28_annotated[df_ASD12_D28_annotated$AB=="A",]
df_B_ASD_D28 <-df_ASD12_D28_annotated[df_ASD12_D28_annotated$AB=="B",]
```

Identify genes mapping in each compartment

```{r}
# (change the df name in the first line of the following code and the output file name, using the df in the previous chunk and listed below) 
#(note that from annont we get only the hgnc symbol but not ensembl names)
#df_A_CTRL_D28
#df_B_CTRL_D28
#df_A_ASD_D28
#df_B_ASD_D28
#df_noswitch_D28_annotated

G <-df_B_ASD_D28$annot.symbol
G = data.frame(G)
#remove NA
G<- filter(G, G != "NA")
#keep only unique terms
G <- unique(G)
#change names to the columns
colnames(G) <- c("hgnc_symbol")
write.table(G, file="GENES_B_ASD_D28.txt", quote=FALSE, row.names = F)

#identify GENES in the switch region (change the df name in the first line and the output file name -df_AtoB_annotated, -df_BtoA_annotated, -df_switch_annotated)
SG <-df_AtoB_annotated$annot.symbol
SG = data.frame(SG)
#remove NA
SG<- filter(SG, SG != "NA")
#keep only unique terms
SG <- unique(SG)
#change names to the columns
colnames(SG) <- c("hgnc_symbol")
write.table(SG, file="GENES_AtoB_switch_regions_D28.txt", quote=FALSE, row.names = F)

DEG_D28<-read.csv("/RNAseq/DEG_ASD12D28_vs_CTRL9D28.csv", sep = ",", row.names = 1)

#select genes in common between the switch region and the DEGfile from DeSeq2-all (run it after the previous line for each of the datasets analyzed -df_AtoB_annotated, -df_BtoA_annotated, -df_switch_annotated)
mergeSG_DEG <- merge(DEG_D28,SG, by="hgnc_symbol", all=F)
#remove NA
mergeSG_DEG<- filter(mergeSG_DEG, padj != "NA")

#select DEG with BaseMean>20 (Expressed)
mergeSG_DEG_exp <- mergeSG_DEG[mergeSG_DEG$baseMean> 20,]
write.csv(mergeSG_DEG_exp, file="expressed_DEG_GENES_AtoB_switch_regions_D28.csv", quote=FALSE, row.names = F)

#read the files generated with the code above
mergeSG_DEG_expAB<- read.csv("/HiC_files/expressed_DEG_GENES_AtoB_switch_regions_D28.csv", header=T)
mergeSG_DEG_expBA<- read.csv("/HiC_files/expressed_DEG_GENES_BtoA_switch_regions_D28.csv", header=T)

#select DEG with significant abs(Log2FC)>0.58 and BaseMean>20 (Expressed)
mergeSG_DEG_expAB_FC <-mergeSG_DEG_expAB[mergeSG_DEG_expAB$log2FoldChange>0.58 | mergeSG_DEG_expAB$log2FoldChange< -0.58,]
mergeSG_DEG_expBA_FC <-mergeSG_DEG_expBA[mergeSG_DEG_expBA$log2FoldChange>0.58 | mergeSG_DEG_expBA$log2FoldChange< -0.58,]


```

Analysis of CONCORDANT GENES (BM>20, |FC|>0.58, overlapping with switch regions)

```{r}
#concordant significant genes (with compartment switch) (AtoB log2FoldChange< -0.58, BtoA log2FoldChange> 0.58)
#Concordant AtoB
c_mergeSG_DEG_exp_AB_FC <- mergeSG_DEG_exp_AB_FC[mergeSG_DEG_exp_AB_FC$log2FoldChange< -0.58,]
write.csv(c_mergeSG_DEG_exp_AB_FC, file="concordantFC_expressed_DEG_GENES_AtoB_switch_regions_D28.csv", quote=FALSE, row.names = F)
#concordant BtoA
c_mergeSG_DEG_exp_BA_FC <- mergeSG_DEG_exp_BA_FC[mergeSG_DEG_exp_BA_FC$log2FoldChange>0.58,]
write.csv(c_mergeSG_DEG_exp_BA_FC, file="concordantFC_expressed_DEG_GENES_BtoA_switch_regions_D28.csv", quote=FALSE, row.names = F)

```

The concordant genes from the last lines of the analysis have been then uploaded to WebGestalt to investigate pathway enrichment.
